client/client.go

```diff
@@ -1,71 +1,378 @@
+			return err
+		}
+	}
+	c.db.AddRole("root", &data.Role{Threshold: 1, KeyIDs: rootKeyIDs})
+
+	if err := c.decodeRoot(rootJSON); err != nil {
+		return err
+	}
+
+	return c.local.SetMeta("root.json", rootJSON)
+}
+
+// Update downloads and verifies remote metadata and returns updated targets.
+//
+// It performs the update part of "The client application" workflow from
+// section 5.1 of the TUF spec:
+//
+// https://github.com/theupdateframework/tuf/blob/v0.9.9/docs/tuf-spec.txt#L714
+func (c *Client) Update() (data.Files, error) {
+	// Always start the update using local metadata
+	if err := c.getLocalMeta(); err != nil {
 		return nil, err
 	}
-	return &Client{repo, remote}, nil
+
+	// TODO: If we get an invalid signature downloading timestamp.json
+	//       or snapshot.json, download root.json and start again.
+
+	// Get timestamp.json, extract snapshot.json file meta and save the
+	// timestamp.json locally
+	timestampJSON, err := c.downloadMeta("timestamp.json", nil)
+	if err != nil {
+		return nil, err
+	}
+	snapshotMeta, err := c.decodeTimestamp(timestampJSON)
+	if err != nil {
+		return nil, err
+	}
+	if err := c.local.SetMeta("timestamp.json", timestampJSON); err != nil {
+		return nil, err
+	}
+
+	// Return ErrLatest if we already have the latest snapshot.json
+	if c.hasMeta("snapshot.json", snapshotMeta) {
+		return nil, ErrLatest
+	}
+
+	// Get snapshot.json, then extract root.json and targets.json file meta.
+	//
+	// The snapshot.json is only saved locally after checking root.json and
+	// targets.json so that it will be re-downloaded on subsequent updates
+	// if this update fails.
+	snapshotJSON, err := c.downloadMeta("snapshot.json", &snapshotMeta)
+	if err != nil {
+		return nil, err
+	}
+	rootMeta, targetsMeta, err := c.decodeSnapshot(snapshotJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we don't have the root.json, download it, save it in local
+	// storage and restart the update
+	if !c.hasMeta("root.json", rootMeta) {
+		rootJSON, err := c.downloadMeta("root.json", &rootMeta)
+		if err != nil {
+			return nil, err
+		}
+		if err := c.decodeRoot(rootJSON); err != nil {
+			return nil, err
+		}
+		if err := c.local.SetMeta("root.json", rootJSON); err != nil {
+			return nil, err
+		}
+		return c.Update()
+	}
+
+	// If we don't have the targets.json, download it, determine updated
+	// targets and save targets.json in local storage
+	var updatedTargets data.Files
+	if !c.hasMeta("targets.json", targetsMeta) {
+		targetsJSON, err := c.downloadMeta("targets.json", &targetsMeta)
+		if err != nil {
+			return nil, err
+		}
+		updatedTargets, err = c.decodeTargets(targetsJSON)
+		if err != nil {
+			return nil, err
+		}
+		if err := c.local.SetMeta("targets.json", targetsJSON); err != nil {
+			return nil, err
+		}
+	}
+
+	// Save the snapshot.json now it has been processed successfully
+	if err := c.local.SetMeta("snapshot.json", snapshotJSON); err != nil {
+		return nil, err
+	}
+
+	return updatedTargets, nil
```

Comment:
So because `Update` always starts from local metadata, there is no need to assign downloaded metadata to the `Client` fields during the update, the downloaded JSON just gets set in local storage once decoded / verified.

The only current use of the client metadata fields is for checking the versions when downloading new metadata, so if the fields seem confusing I could replace them with say `localRootVersion`, `localTargetsVersion` etc.

WDYT?